File: nilebyte_clinic/nilebyte_clinic/page/clinic_admin_dashboard/clinic_admin_dashboard.js
frappe.pages["clinic-admin-dashboard"].on_page_load = function (wrapper) {
  new NBClinicAdminDashboard(wrapper);
};

class NBClinicAdminDashboard {
  constructor(wrapper) {
    this.wrapper = wrapper;
    this.page = frappe.ui.make_app_page({
      parent: wrapper,
      title: "Clinic Admin Dashboard",
      single_column: true,
    });

    this.method_base = "nilebyte_clinic.nilebyte_clinic.page.clinic_admin_dashboard.clinic_admin_dashboard";
    this.dashboard = {};
    this.search_timer = null;

    this.render();
    this.bind_events();
    this.load_dashboard();
    this.refresh_interval = setInterval(() => this.load_dashboard(false), 60000);
  }

  render() {
    $(this.page.body).html(`
      <div class="nb-clinic-app">
        <aside class="nb-sidebar">
          <div class="nb-brand">
            <div class="nb-brand-logo">✦</div>
            <div class="nb-brand-text">
              <div class="nb-brand-title">ClinicWiser</div>
              <div class="nb-brand-sub">Admin Portal</div>
            </div>
          </div>
          <nav class="nb-nav">
            <button class="nb-nav-item active" data-route="clinic-admin-dashboard">
              <span class="nb-nav-icon">⌂</span>
              <span class="nb-nav-text">Dashboard</span>
            </button>
            <button class="nb-nav-item" data-route="clinic-reports">
              <span class="nb-nav-icon">▥</span>
              <span class="nb-nav-text">Reports</span>
            </button>
            <button class="nb-nav-item" data-route="List/Healthcare Practitioner">
              <span class="nb-nav-icon">👥</span>
              <span class="nb-nav-text">Staff</span>
            </button>
            <button class="nb-nav-item" data-route="List/Item">
              <span class="nb-nav-icon">📦</span>
              <span class="nb-nav-text">Inventory</span>
            </button>
            <button class="nb-nav-item" data-route="List/Sales Invoice">
              <span class="nb-nav-icon">💳</span>
              <span class="nb-nav-text">Billing</span>
            </button>
            <button class="nb-nav-item" data-action="settings">
              <span class="nb-nav-icon">⚙</span>
              <span class="nb-nav-text">Settings</span>
            </button>
          </nav>
        </aside>

        <main class="nb-main">
          <header class="nb-topbar">
            <div class="nb-page-title">
              <h1>Clinic Administration</h1>
              <p>Global Performance Overview</p>
            </div>
            <div class="nb-top-actions">
              <div class="nb-search-bar">
                <span class="nb-search-icon">⌕</span>
                <input id="nb-admin-search" placeholder="Search records..." autocomplete="off" />
                <div id="nb-admin-search-results" class="nb-admin-search-results"></div>
              </div>
              <div class="nb-user-profile">
                <div class="nb-avatar">AD</div>
              </div>
            </div>
          </header>

          <div class="nb-content nb-fade-in">
            <div id="nb-admin-kpis" class="nb-kpis"></div>

            <div class="nb-grid nb-grid-2">
              <section class="nb-card">
                <div class="nb-card-header">
                  <h3>Financial Overview</h3>
                  <div class="nb-tabs" style="padding: 0;">
                    <button class="active" data-chart="revenue">Revenue</button>
                    <button data-chart="expenses">Expenses</button>
                  </div>
                </div>
                <div id="nb-admin-chart" style="height: 300px; padding: 20px;">
                  <div class="nb-empty-soft">Loading chart...</div>
                </div>
              </section>

              <section class="nb-card">
                <div class="nb-card-header"><h3>Facility Status</h3></div>
                <div id="nb-admin-facility" style="padding: 20px;"></div>
              </section>
            </div>

            <div id="nb-admin-quick-cards" class="nb-grid nb-grid-3"></div>

            <div class="nb-grid nb-grid-3">
              <section class="nb-card">
                <div class="nb-card-header"><h3>Recent Activities</h3></div>
                <div id="nb-admin-activities" class="nb-table-container"></div>
              </section>
              <section class="nb-card">
                <div class="nb-card-header"><h3>Active Tasks</h3></div>
                <div id="nb-admin-tasks" class="nb-table-container"></div>
              </section>
              <section class="nb-card">
                <div class="nb-card-header"><h3>System Status</h3></div>
                <div id="nb-admin-settings" class="nb-table-container"></div>
              </section>
            </div>
          </div>
        </main>
      </div>
    `);
  }

  bind_events() {
    $(this.wrapper).on("click", ".nb-nav-item", (e) => {
      const $btn = $(e.currentTarget);
      $(".nb-nav-item").removeClass("active");
      $btn.addClass("active");
      const route = $btn.data("route");
      const action = $btn.data("action");
      if (route) this.go_route(route);
      if (action === "settings") this.open_settings();
    });

    $(this.wrapper).on("click", "[data-admin-route]", (e) => {
      const route = $(e.currentTarget).attr("data-admin-route");
      this.go_route(route);
    });

    $("#nb-admin-search").on("input", () => {
      clearTimeout(this.search_timer);
      const query = $("#nb-admin-search").val().trim();
      this.search_timer = setTimeout(() => this.search_global(query), 250);
    });

    $(document).on("click", function (e) {
      if (!$(e.target).closest(".nb-search-bar").length) {
        $("#nb-admin-search-results").hide();
      }
    });
  }

  call(method, args, callback) {
    frappe.call({
      method: `${this.method_base}.${method}`,
      args: args || {},
      freeze: false,
      callback: (r) => callback && callback(r.message),
      error: () => frappe.show_alert({ message: `Failed to load ${method}`, indicator: "red" }),
    });
  }

  load_dashboard() {
    this.call("get_admin_dashboard", { days: 30 }, (data) => {
      this.dashboard = data || {};
      this.render_kpis(this.dashboard.kpis || {});
      this.render_chart(this.dashboard.chart || {});
      this.render_facility(this.dashboard.facility || {});
      this.render_quick_cards(this.dashboard.quick_cards || []);
      this.render_list("#nb-admin-activities", this.dashboard.recent_activities || []);
      this.render_list("#nb-admin-settings", this.dashboard.settings || []);
      this.render_list("#nb-admin-tasks", this.dashboard.tasks || []);
    });
  }

  render_kpis(k) {
    const cards = [
      { label: "Total Patients", value: this.num(k.total_patients), icon: "👤", color: "blue", route: "List/Patient" },
      { label: "Revenue", value: `EGP ${this.money(k.monthly_revenue)}`, icon: "💰", color: "red", route: "query-report/Clinic Revenue Summary" },
      { label: "Pending Appts", value: this.num(k.pending_appointments), icon: "📅", color: "green", route: "List/Patient Appointment" },
      { label: "Low Stock", value: this.num(k.low_stock_items), icon: "📦", color: "orange", route: "List/Bin" },
    ];
    $("#nb-admin-kpis").html(cards.map(c => `
      <div class="nb-kpi-card" data-admin-route="${this.esc(c.route)}">
        <div class="nb-kpi-icon" style="background:var(--nb-primary)">${c.icon}</div>
        <div class="nb-kpi-info">
            <h4>${c.label}</h4>
            <div class="nb-kpi-value">${this.esc(c.value)}</div>
        </div>
      </div>
    `).join(""));
  }

  render_chart(chart) {
    const labels = chart.labels || [];
    const revenue = chart.revenue || [];
    const expenses = chart.expenses || [];
    const max = Math.max(1, ...revenue, ...expenses);
    if (!labels.length) {
      $("#nb-admin-chart").html(`<div class="nb-empty-soft">No financial data found.</div>`);
      return;
    }
    const html = `<div style="display: flex; align-items: flex-end; height: 200px; gap: 10px; border-bottom: 1px solid var(--nb-border);">
        ${labels.map((label, i) => {
            const rh = Math.max(5, (revenue[i] / max) * 100);
            const eh = Math.max(5, (expenses[i] / max) * 100);
            return `
                <div style="flex: 1; display: flex; flex-direction: column; align-items: center; gap: 5px;">
                    <div style="display: flex; gap: 2px; align-items: flex-end; height: 100px;">
                        <div style="width: 10px; height: ${rh}%; background: var(--nb-primary); border-radius: 2px 2px 0 0;"></div>
                        <div style="width: 10px; height: ${eh}%; background: var(--nb-danger); border-radius: 2px 2px 0 0;"></div>
                    </div>
                    <span style="font-size: 10px; color: var(--nb-text-muted)">${label}</span>
                </div>
            `;
        }).join("")}
    </div>`;
    $("#nb-admin-chart").html(html);
  }

  render_facility(facility) {
    $("#nb-admin-facility").html(`
      <div style="display: grid; gap: 15px;">
        <div style="display: flex; justify-content: space-between; padding: 12px; background: var(--nb-surface); border-radius: 8px;">
            <span style="font-weight: 600;">Active Doctors</span>
            <b style="font-size: 18px;">${this.num(facility.active_doctors)}</b>
        </div>
        <div style="display: flex; justify-content: space-between; padding: 12px; background: var(--nb-surface); border-radius: 8px;">
            <span style="font-weight: 600;">Service Units</span>
            <b style="font-size: 18px;">${this.num(facility.rooms)}</b>
        </div>
        <div style="display: flex; justify-content: space-between; padding: 12px; background: var(--nb-surface); border-radius: 8px;">
            <span style="font-weight: 600;">Today Admissions</span>
            <b style="font-size: 18px;">${this.num(facility.today_admissions)}</b>
        </div>
      </div>
    `);
  }

  render_quick_cards(cards) {
    $("#nb-admin-quick-cards").html(cards.map((c, i) => `
      <section class="nb-card" style="display: flex; align-items: center; justify-content: space-between; padding: 20px;">
        <div style="font-weight: 800; font-family: 'Outfit'; font-size: 16px;">${this.esc(c.label)}</div>
        <button class="nb-btn nb-btn-primary" data-admin-route="${this.esc(this.route_string(c.route))}">${this.esc(c.button || "Open")}</button>
      </section>
    `).join(""));
  }

  render_list(selector, rows) {
    if (!rows.length) {
      $(selector).html(`<div class="nb-empty-soft">No records found.</div>`);
      return;
    }
    $(selector).html(`
        <div style="padding: 10px;">
            ${rows.map(row => `
                <div data-admin-route="${this.esc(this.route_string(row.route))}" 
                     style="display: flex; align-items: center; gap: 12px; padding: 12px; border-bottom: 1px solid var(--nb-border); cursor: pointer;">
                    <div class="nb-avatar" style="width: 32px; height: 32px; font-size: 14px;">${this.initials(row.label || row.title)}</div>
                    <div style="flex: 1;">
                        <div style="font-weight: 700;">${this.esc(row.label || row.title)}</div>
                        <div style="font-size: 11px; color: var(--nb-text-muted)">${this.esc(row.subtitle || "")}</div>
                    </div>
                </div>
            `).join("")}
        </div>
    `);
  }

  search_global(query) {
    if (!query || query.length < 2) {
      $("#nb-admin-search-results").hide();
      return;
    }
    this.call("search_global", { query }, (rows) => {
      rows = rows || [];
      if (!rows.length) {
        $("#nb-admin-search-results").html(`<div style="padding: 12px;">No results</div>`).show();
        return;
      }
      $("#nb-admin-search-results").html(rows.map(row => `
        <div class="nb-global-item" data-admin-route="${this.esc(this.route_string(row.route))}" style="padding: 12px; border-bottom: 1px solid var(--nb-border); cursor: pointer;">
          <b style="display:block;">${this.esc(row.title || row.name)}</b>
          <div style="font-size: 11px; color: var(--nb-text-muted)">${this.esc(row.type || "")}</div>
        </div>
      `).join("")).show();
    });
  }

  open_settings() {
    this.go_route("Form/NileByte Clinic Settings/NileByte Clinic Settings");
  }

  go_route(route) {
    if (!route) return;
    const parts = String(route).split("/").map(decodeURIComponent);
    frappe.set_route(...parts);
  }

  route_string(route) {
    if (!route) return "clinic-admin-dashboard";
    if (Array.isArray(route)) return route.map(v => encodeURIComponent(v)).join("/");
    return String(route);
  }

  initials(value) {
    const words = String(value || "A").trim().split(/\s+/).slice(0, 2);
    return words.map(w => w.charAt(0).toUpperCase()).join("") || "A";
  }

  num(value) {
    return Number(value || 0).toLocaleString();
  }

  money(value) {
    return Number(value || 0).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
  }

  esc(value) {
    return frappe.utils.escape_html(String(value == null ? "" : value));
  }
}

File: nilebyte_clinic/nilebyte_clinic/page/clinic_admin_dashboard/clinic_admin_dashboard.py
# Copyright (c) 2026, NileByte
# For license information, please see license.txt

"""Clinic Admin Dashboard backend for NileByte Clinic.

Route: /app/clinic-admin-dashboard

The implementation is defensive so it works across ERPNext Healthcare/Marley
variants. Every public method checks role access and every optional DocType or
field is inspected before use.
"""

import frappe
from frappe.utils import nowdate, add_days, getdate, flt, cint, date_diff

ALLOWED_ROLES = ("Clinic Admin", "System Manager")


def get_context(context):
    _ensure_page_access()
    context.no_cache = 1
    context.title = "Clinic Admin Dashboard"


def _roles():
    if frappe.session.user == "Administrator":
        return {"Administrator", "System Manager", "Clinic Admin"}
    return set(frappe.get_roles(frappe.session.user))


def _ensure_page_access():
    if frappe.session.user == "Administrator":
        return
    if not _roles().intersection(set(ALLOWED_ROLES)):
        frappe.throw("You do not have permission to access Clinic Admin Dashboard.", frappe.PermissionError)


def has_doctype(doctype):
    return bool(frappe.db.exists("DocType", doctype))


def has_field(doctype, fieldname):
    return bool(has_doctype(doctype) and frappe.db.has_column(doctype, fieldname))


def _money(value):
    return flt(value or 0, 2)


def _date_range(days=30):
    return add_days(nowdate(), -abs(cint(days)) + 1), nowdate()


def _normalize_status(status):
    if status in ["Scheduled", "Open", "Waiting"]:
        return "Waiting"
    if status in ["Checked In", "In Progress", "With Doctor"]:
        return "With Doctor"
    if status in ["Closed", "Completed"]:
        return "Completed"
    if status in ["Cancelled"]:
        return "Cancelled"
    if status in ["No Show", "No-show"]:
        return "No Show"
    return status or "Waiting"


@frappe.whitelist()
def get_admin_dashboard(days=30):
    _ensure_page_access()
    return {
        "kpis": get_kpis(days),
        "chart": get_chart_data(days),
        "facility": get_facility_stats(),
        "tasks": get_admin_tasks(),
        "recent_activities": get_recent_activities(),
        "settings": get_settings_overview(),
        "reports": get_report_cards(),
        "quick_cards": get_quick_cards(),
    }


@frappe.whitelist()
def get_kpis(days=30):
    _ensure_page_access()
    start, end = _date_range(days)
    today = nowdate()

    total_patients = frappe.db.count("Patient") if has_doctype("Patient") else 0

    monthly_revenue = 0
    revenue_today = 0
    unpaid_count = 0
    unpaid_amount = 0
    if has_doctype("Sales Invoice"):
        monthly_revenue = frappe.db.sql(
            """
            SELECT COALESCE(SUM(base_grand_total), SUM(grand_total), 0)
            FROM `tabSales Invoice`
            WHERE docstatus = 1 AND posting_date BETWEEN %s AND %s
            """,
            (start, end),
        )[0][0] or 0
        revenue_today = frappe.db.sql(
            """
            SELECT COALESCE(SUM(base_grand_total), SUM(grand_total), 0)
            FROM `tabSales Invoice`
            WHERE docstatus = 1 AND posting_date = %s
            """,
            today,
        )[0][0] or 0
        unpaid = frappe.db.sql(
            """
            SELECT COUNT(*), COALESCE(SUM(outstanding_amount), 0)
            FROM `tabSales Invoice`
            WHERE docstatus = 1 AND outstanding_amount > 0
            """
        )[0]
        unpaid_count = unpaid[0] or 0
        unpaid_amount = unpaid[1] or 0

    pending_appointments = 0
    completed_today = 0
    no_show_today = 0
    appointments_today = 0
    if has_doctype("Patient Appointment"):
        rows = frappe.get_all(
            "Patient Appointment",
            filters={"appointment_date": today},
            fields=["status"],
        )
        appointments_today = len(rows)
        mapped = [_normalize_status(r.status) for r in rows]
        pending_appointments = mapped.count("Waiting") + mapped.count("With Doctor")
        completed_today = mapped.count("Completed")
        no_show_today = mapped.count("No Show")

    active_doctors = frappe.db.count("Healthcare Practitioner") if has_doctype("Healthcare Practitioner") else 0
    no_show_rate = round((no_show_today / appointments_today) * 100, 1) if appointments_today else 0

    low_stock_items = 0
    if has_doctype("Bin"):
        low_stock_items = frappe.db.sql(
            """
            SELECT COUNT(DISTINCT item_code)
            FROM `tabBin`
            WHERE actual_qty <= 0
            """
        )[0][0] or 0

    return {
        "total_patients": total_patients,
        "monthly_revenue": _money(monthly_revenue),
        "revenue_today": _money(revenue_today),
        "pending_appointments": pending_appointments,
        "low_stock_items": low_stock_items,
        "appointments_today": appointments_today,
        "completed_today": completed_today,
        "unpaid_invoices": unpaid_count,
        "unpaid_amount": _money(unpaid_amount),
        "active_doctors": active_doctors,
        "no_show_rate": no_show_rate,
    }


@frappe.whitelist()
def get_chart_data(days=7):
    _ensure_page_access()
    start, end = _date_range(days)
    labels = []
    revenue = []
    expenses = []

    current = getdate(start)
    end_date = getdate(end)
    while current <= end_date:
        label = current.strftime("%d %b")
        day = current.isoformat()
        labels.append(label)
        revenue.append(_money(_invoice_total(day)))
        expenses.append(_money(_expense_total(day)))
        current = getdate(add_days(current, 1))

    return {"labels": labels, "revenue": revenue, "expenses": expenses}


def _invoice_total(day):
    if not has_doctype("Sales Invoice"):
        return 0
    return frappe.db.sql(
        """
        SELECT COALESCE(SUM(base_grand_total), SUM(grand_total), 0)
        FROM `tabSales Invoice`
        WHERE docstatus = 1 AND posting_date = %s
        """,
        day,
    )[0][0] or 0


def _expense_total(day):
    if not has_doctype("Purchase Invoice"):
        return 0
    return frappe.db.sql(
        """
        SELECT COALESCE(SUM(base_grand_total), SUM(grand_total), 0)
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1 AND posting_date = %s
        """,
        day,
    )[0][0] or 0


@frappe.whitelist()
def get_facility_stats():
    _ensure_page_access()
    patient = _recent_patient()
    return {
        "patient": patient,
        "active_doctors": frappe.db.count("Healthcare Practitioner") if has_doctype("Healthcare Practitioner") else 0,
        "rooms": frappe.db.count("Healthcare Service Unit") if has_doctype("Healthcare Service Unit") else 0,
        "today_admissions": frappe.db.count("Inpatient Record", {"admission_date": nowdate()}) if has_doctype("Inpatient Record") and has_field("Inpatient Record", "admission_date") else 0,
    }


def _recent_patient():
    if not has_doctype("Patient"):
        return None
    fields = ["name"]
    for field in ["patient_name", "first_name", "sex", "dob", "mobile", "phone"]:
        if has_field("Patient", field):
            fields.append(field)
    rows = frappe.get_all("Patient", fields=fields, order_by="modified desc", limit=1)
    if not rows:
        return None
    row = rows[0]
    name = row.get("patient_name") or row.get("first_name") or row.get("name")
    row["display_name"] = name
    row["age_gender"] = _age_gender(row)
    row["conditions"] = _patient_conditions(row.get("name"))
    return row


def _age_gender(row):
    parts = []
    if row.get("dob"):
        try:
            today = getdate(nowdate())
            born = getdate(row.get("dob"))
            years = today.year - born.year - ((today.month, today.day) < (born.month, born.day))
            if years >= 0:
                parts.append(f"{years} Years")
        except Exception:
            pass
    if row.get("sex"):
        parts.append(row.get("sex"))
    return ", ".join(parts)


def _patient_conditions(patient):
    if not patient:
        return ""
    if has_doctype("Patient Medical Record") and has_field("Patient Medical Record", "patient"):
        rows = frappe.get_all(
            "Patient Medical Record",
            filters={"patient": patient},
            fields=["subject" if has_field("Patient Medical Record", "subject") else "name"],
            limit=2,
            order_by="creation desc",
        )
        values = [r.get("subject") or r.get("name") for r in rows]
        return ", ".join([v for v in values if v])
    return ""


@frappe.whitelist()
def get_admin_tasks():
    _ensure_page_access()
    tasks = []
    if has_doctype("ToDo"):
        rows = frappe.get_all(
            "ToDo",
            filters={"allocated_to": frappe.session.user, "status": ["!=", "Closed"]},
            fields=["name", "description", "priority", "date"],
            order_by="modified desc",
            limit=5,
        )
        for row in rows:
            tasks.append({
                "label": row.description or row.name,
                "status": row.priority or "Open",
                "route": ["Form", "ToDo", row.name],
            })
    if not tasks:
        tasks = [
            {"label": "Review unpaid invoices", "status": "Due Soon", "route": ["List", "Sales Invoice"]},
            {"label": "Review pending appointments", "status": "Open", "route": ["List", "Patient Appointment"]},
            {"label": "Check user roles", "status": "Admin", "route": ["List", "User"]},
        ]
    return tasks


@frappe.whitelist()
def get_recent_activities():
    _ensure_page_access()
    activities = []
    if has_doctype("Version"):
        rows = frappe.get_all(
            "Version",
            fields=["ref_doctype", "docname", "owner", "modified"],
            order_by="modified desc",
            limit=5,
        )
        for row in rows:
            activities.append({
                "title": f"{row.owner} updated {row.ref_doctype}",
                "subtitle": row.docname,
                "time": str(row.modified),
                "route": ["Form", row.ref_doctype, row.docname],
            })
    if not activities:
        activities = [
            {"title": "System ready", "subtitle": "No recent activity found", "time": "Now", "route": ["clinic-admin-dashboard"]},
        ]
    return activities


@frappe.whitelist()
def get_settings_overview():
    _ensure_page_access()
    return [
        {"label": "User Roles", "icon": "👥", "route": ["List", "Role"]},
        {"label": "Clinic Settings", "icon": "🏥", "route": _settings_route()},
        {"label": "Billing Setup", "icon": "💳", "route": ["List", "Mode of Payment"]},
        {"label": "Services & Prices", "icon": "🏷", "route": ["List", "Item"]},
        {"label": "Backup Data", "icon": "☁", "route": ["List", "Data Export"]},
    ]


def _settings_route():
    if has_doctype("NileByte Clinic Settings"):
        # Single DocType route is usually Form/Single DocType/Single DocType name.
        return ["Form", "NileByte Clinic Settings", "NileByte Clinic Settings"]
    return ["List", "System Settings"]


@frappe.whitelist()
def get_report_cards():
    _ensure_page_access()
    reports = [
        {"label": "Revenue Reports", "icon": "💹", "route": ["query-report", "Clinic Revenue Summary"]},
        {"label": "Appointment Reports", "icon": "📅", "route": ["query-report", "Appointment Status Report"]},
        {"label": "Unpaid Invoices", "icon": "🧾", "route": ["query-report", "Unpaid Invoices Report"]},
        {"label": "Doctor Performance", "icon": "👨‍⚕️", "route": ["query-report", "Doctor Performance Report"]},
        {"label": "Patient Visit History", "icon": "📁", "route": ["query-report", "Patient Visit History"]},
    ]
    return reports


@frappe.whitelist()
def get_quick_cards():
    _ensure_page_access()
    return [
        {"label": "Revenue", "value": "Open report", "button": "Revenue", "route": ["query-report", "Clinic Revenue Summary"]},
        {"label": "Active Doctors", "value": frappe.db.count("Healthcare Practitioner") if has_doctype("Healthcare Practitioner") else 0, "button": "Manage Doctors", "route": ["List", "Healthcare Practitioner"]},
        {"label": "Inpatient Admissions", "value": frappe.db.count("Inpatient Record") if has_doctype("Inpatient Record") else 0, "button": "Open Admissions", "route": ["List", "Inpatient Record"]},
    ]


@frappe.whitelist()
def search_global(query):
    _ensure_page_access()
    query = (query or "").strip()
    if len(query) < 2:
        return []
    results = []

    if has_doctype("Patient"):
        conditions = []
        params = {"q": f"%{query}%"}
        for field in ["name", "patient_name", "first_name", "mobile", "phone"]:
            if has_field("Patient", field):
                conditions.append(f"{field} LIKE %(q)s")
        if conditions:
            optional_fields = [field for field in ["patient_name", "first_name", "mobile", "phone"] if has_field("Patient", field)]
            select_fields = ["name"] + optional_fields
            sql = "SELECT {0} FROM `tabPatient` WHERE {1} ORDER BY modified DESC LIMIT 8".format(", ".join(select_fields), " OR ".join(conditions))
            rows = frappe.db.sql(sql, params, as_dict=True)
            for row in rows:
                results.append({
                    "type": "Patient",
                    "name": row.name,
                    "title": row.get("patient_name") or row.get("first_name") or row.name,
                    "subtitle": row.get("mobile") or row.get("phone") or row.name,
                    "route": ["Form", "Patient", row.name],
                })

    if has_doctype("Sales Invoice"):
        invs = frappe.get_all(
            "Sales Invoice",
            filters={"name": ["like", f"%{query}%"]},
            fields=["name", "customer", "grand_total", "outstanding_amount"],
            limit=5,
            order_by="modified desc",
        )
        for row in invs:
            results.append({
                "type": "Invoice",
                "name": row.name,
                "title": row.name,
                "subtitle": f"{row.customer or ''} • Outstanding {flt(row.outstanding_amount or 0, 2)}",
                "route": ["Form", "Sales Invoice", row.name],
            })
    return results[:12]
